sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala

```diff
@@ -90,7 +90,7 @@ abstract class UserDefinedType[UserType >: Null] extends DataType with Serializa
   override def hashCode(): Int = getClass.hashCode()
 
   override def equals(other: Any): Boolean = other match {
-    case that: UserDefinedType[_] => this.acceptsType(that)
+    case that: UserDefinedType[_] => this.getClass == other.getClass
```

Comment:
This update comes from https://github.com/apache/spark/commit/0ec17c989debddb56a8048e4bd29ae666e4d9c56. I'm currently checking the python test results.

Reply:
Ah, ok. `branch-3.0` has [the same issue](https://github.com/apache/spark/runs/1307296057), but `test_infer_schema_with_udt` in `branch-3.0` does not fail because it drops the `labeled_point` view created in L350 before running the second `createOrReplaceTempView` in L361:
https://github.com/apache/spark/blob/22392be4315900346ac884ad32bd6332ffc3e64e/python/pyspark/sql/tests/test_types.py#L349-L363

A root cause of this issue is that, since `PythonUserDefinedType#userClass`is always null, `isAssignableFrom` in `UserDefinedType#acceptsType ` throws a null exception. To fix this issue, I think we need to backport https://issues.apache.org/jira/browse/SPARK-32090 into branch-3.0/2.4. WDYT? @dongjoon-hyun @viirya @HyukjinKwon

Reply:
+1 for backporting. Thank you for analysis, @maropu .

Reply:
Looks okay. Thanks for looking at it.

Reply:
okay, I will. Thanks for the check, @dongjoon-hyun and @viirya !

Reply:
done, thanks, all!